### PR TITLE
feat(watermarker): support JS library export

### LINKED ISSUE #24: Make watermarker library usable as JavaScript library
# 🚀 Feature Request

## Current Problem
The watermarker library is usable in Java and Kotlin. It is possible to use the library in the browser by using a Kotlin frontend like KVision. However, it is currently not possible to import the library into plain JavaScript.

## Proposed Solution

Evaluate what is required to export the library as JavaScript library

## Additional Context
More information: https://kotlinlang.org/docs/js-to-kotlin-interop.html

### PR DESCRIPTION
**Pull request by gemdav**
## Description

Enables export of the library as a JS library. After further investigation, no better solutions than the one already described in [this comment](https://github.com/FraunhoferISST/TREND/issues/24#issuecomment-2091191097) of issue  #24 have been found. 

## Linked Issue(s)

Fixes #24

## CLA
By submitting this pull request, I have read the [Corporate Contributor License Agreement (CLA)](https://github.com/FraunhoferISST/TREND/blob/main/CLA.md) and I hereby accept and sign the CLA.

### CODE REVIEW

**Comment by hnorkowski** [comment]
This looks promising, but I have several points to discuss before we merge this:

- [Solved] we should add a description how to export and use the library in JS
  -  plain JS
  - NodeJS
  - ...?
- [Solved] we should OptIn into JS export for the whole project so we do not have to write it before every `@JsExport`
  - I already did this locally, maybe I can push this to this PR
- [Solved] you cannot annotate subclasses with `@JsExport`: are those available as soon as `@JsExport` is annoted to the superclass?
  - Yes. They are exported at the same namespace as the superclass. Currently this works fine but it could lead to clashes in the future.
- [Solved] `binaries.library()` should be changed to `binaries.executable()` so that the library is usable from plain javascript and not only as node module.
- [Moved to Issue #44] There is an endless list of warnings like these:
`Exported declaration uses non-exportable parameter type: List<Byte>`
  - Some of the types we use are not supported in JS (e.g., `Byte`, `Char`)
  - The code still works but you cannot interact directly with such types, you can only pass them to other kotlin functions that require them.
  - Maybe we can reduce the amount of types that are unusable?
   - [Solved] e.g., #38 can be solved by changing the type of placement from `Sequence<Int>` (unsupported by JS) to `List<Int>` (supported by JS)

**Comment by hnorkowski** [comment]
I created an issue regarding the public facing types that are not supported by JS (see #44). It requires a lot of work and is out-of-scope for this PR.

Therefore, I think this PR is ready.
Because I contributed to it as well it should be reviewed by an additional maintainer.

**Review by hnorkowski — COMMENTED** [review]
I created an issue regarding the public facing types that are not supported by JS (see https://github.com/FraunhoferISST/TREND/issues/44). It requires a lot of work and is out-of-scope for this PR.

Therefore, I think this PR is ready.
Because I contributed to it as well it should be reviewed by an additional maintainer.

File: watermarker/src/commonMain/kotlin/fileWatermarker/TextWatermarker.kt
Comment (gemdav) [inline watermarker/src/commonMain/kotlin/fileWatermarker/TextWatermarker.kt]: Just one small remark from my side.

`[Sequence].toList()` as it is used now for the placement property of the TextWatermarker class usually returns an `ArrayList` type which is nice to handle using JS as it is parseable to a JS array using `[ArrayList].toArray()`. However, if the placement is an empty list, it is not an empty `ArrayList`, but of type `EmptyList` (which is Kotlin-ish i guess, so I did not change it). However, this makes it a bit more inconvenient to work with in JS as the `EmptyList` type does not support the `.toArray()` function call.

Currently, when trying to get the placement as a JS array, instead of

```Typescript
return marker.placement(text).toArray();
```

you have to do something like

```Typescript
let placement = marker.placement(text);
if (typeof placement.toArray !== 'function') {
    return []; // .toArray() is not available on Kotlins EmptyList type, so return [] (empty array) if it is unavailable on the placement object
}
return placement.toArray();
```

or 

```Typescript
try {
    return marker.placement(text).toArray();
} catch (err) {
    if (err instanceof TypeError) {
        return []; // .toArray() is not available on Kotlins EmptyList type, so return [] (empty array) on a TypeError
    }
    throw err;
}
```

this works fine, but 
1. you have to know this
2. it is a bit inconvenient even if you know it.

This can be "solved" using e.g. `.toCollection(ArrayList())` instead of `.toList()` here which returns an `ArrayList` even for empty sequences. I am no Kotlin expert, but I assume that this `EmptyList` type is just the Kotlin way to do it (e.g. to do empty checks by type like `if (list is EmptyList)`), so it's maybe better to leave it like it is.

File: samples/js_modules/js_modules_minimal_example.js
Comment (mhellmeier) [inline samples/js_modules/js_modules_minimal_example.js:7]: Please add a license header to this file.

**Review by mhellmeier — CHANGES_REQUESTED** [review]

File: watermarker/src/commonMain/kotlin/fileWatermarker/TextWatermarker.kt
Comment (hnorkowski) [reply watermarker/src/commonMain/kotlin/fileWatermarker/TextWatermarker.kt]: Good catch, fixed for now with `toMutableList()`. However, this is suspect to future work, see #44.

**Review by mhellmeier — APPROVED** [review]

**Pull request merged**
